### ISSUE TITLE
Git Generator directories in ApplicationSet Helm based application not working

### ISSUE THREAD

**Issue by pratikkava1189** [argoproj/argo-cd#20314]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

I'm using git generator in helm based applicationset

~~~
directory tree
|-- base
|	 `-- root-app
|			|-- charts.yaml
|			|-- templates
|			|  `-- baremetal-applicationset-cluster.yaml
|			|-- values-lab.yaml
`-- clusters
	`-- configs
		|-- lab
		|	`-- app1
		|		|-- deploy.yaml
		|  		|-- kustomization.yaml
		|	`-- app2
		|		|-- deploy.yaml
		|  		|-- kustomization.yaml
~~~

~~~
values.yaml file content
      repoURL: https://github.com/example.git
      targetRevision: gitgen
~~~

~~~
root-app.yaml file content
      apiVersion: argoproj.io/v1alpha1
      kind: ApplicationSet
      metadata:
        name: root-applicationset #<-- looks like this path is taking and making conflicts with git/directory/path in '{{ .path }}
        namespace: argocd
      spec:
        generators:
          - git:
              repoURL: {{ .Values.repoURL }}    # Pull from Helm values-lab.yaml or overrides
              revision: {{ .Values.targetRevision }} 
              directories:
                - path: clusters/configs/{{ .Values.environment }}/* # example lab <--- This path is not passing in '{{ .path }}'
        template:
          metadata:
            name: '{{ .path.basename }}'  # Extract directory name as app name # APP - app1 & app2
          spec:
            project: default
            source:
              repoURL: {{ .Values.repoURL }}                 # Same repo
              targetRevision: {{ .Values.targetRevision }}         # Same revision
              path: {{ .path.path }} # Path to the config
              kustomize: {}
            destination:
              server: 'https://kubernetes.default.svc'
              namespace: '{{ .path.basename }}'  # Use directory name as namespace
            syncPolicy:
              automated:
                prune: true
                selfHeal: true
~~~

**To Reproduce**

see above appset

**Expected behavior**

should be generate dynamic application with git directories as mentioned [here](https://argo-cd.readthedocs.io/en/latest/operator-manual/applicationset/Generators-Git/#git-generator-directories)

**Screenshots**

```shell
v2.10.3
```

**Logs**

```
Failed to load target state: failed to generate manifest for source 1 of 1 rpc error= Manifest generation error (cache), failed exist status 1: Error templating root-app.yaml at <.path.basename>: nill pointer evaluting interface {}.basename
```

**Comment by rumstead** [comment]
It looks like you are using go templating based on `{{ .path.basename}}`. Do you have `goTemplate: true` in your appset? 

Additionally, if you are looking to use a helm chart to template out your applicationsets you will need to escape the helm templates as described [here](https://argo-cd.readthedocs.io/en/latest/operator-manual/applicationset/Template/#deploying-applicationset-resources-as-part-of-a-helm-chart).

**Comment by tomcius** [comment]
Have you found the root cause ? 
I am facing exactly the same issue. I am following the ApplicationSet example but it doesn't work.

`at <.path.basename>: nil pointer evaluating interface {}.basename`

Any idea ?

**Comment by jrobertson-insite** [comment]
I ran into this and the comments here helped me figure out my issue. I'll post an example for good measure. 

TLDR; If you are using helm templating for the ApplicationSet you need to escape the templating in the ApplicationSet since both templating engines use the format `{{ variable/method/whatever }}`.

This is in Argo CD's examples
```
    metadata:
      name: '{{.path.basename}}'
```

But if you use helm templating within Argo CD and load the ApplicationSet via one of those Helm charts you have to do this. (for each of the generator template variable usages)
```
    metadata:
      name: '{{`{{.path.basename}}`}}'
```
